streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1760,15 +1764,14 @@ object ZStream extends ZStreamPlatformSpecific {
     halt(Cause.fail(error))
 
   /**
-   * Creates a stream that emits no elements, never fails and executes
-   * the finalizer before it ends.
+   * Creates an empty stream that never fails and executes the finalizer before it ends.
    */
   final def finalizer[R](finalizer: ZIO[R, Nothing, _]): ZStream[R, Nothing, Nothing] =
     new ZStream[R, Nothing, Nothing] {
-      def fold[R1 <: R, E1, A1, S]: ZStream.Fold[R1, E1, A1, S] =
-        ZManaged.succeed { (s, _, _) =>
-          ZManaged.reserve(Reservation(UIO.succeed(s), _ => finalizer))
-        }
+      def fold[R1 <: R, E, A1, S]: Fold[R1, E, A1, S] = foldDefault
+
+      override def process: ZManaged[R, Nothing, InputStream[Nothing, Nothing]] =
+        ZManaged.succeed(IO.fail(None)).ensuring(finalizer)
```

Comment:
Let's introduce a companion object for `InputStream` with common constructors:
```
object InputStream {
  val end = IO.fail(None)
  def emit[A](a: A) = ...
  def fail[E](e: E) = ...
  def halt[E](c: Cause[E]) = ...
  def die(t: Throwable) = ...
  def dieMessage(m: String) = ...
  def done[E, A](e: Exit[E, A])  = ...
}
```

Reply:
It lives directly inside ZStream companion object alongside the type alias right?

Reply:
Yeah.

Reply:
`halt` and `done` actually need `Option[E]` as the error type parameter of the `Cause` and `Exit` respectively. What should I do?

Reply:
`halt = ZIO.halt(c).mapError(Some(_)), done = ZIO.done(e).mapError(Some(_))`

Reply:
Done. Thanks.